File: backend/models/Message.js
const mongoose = require('mongoose');

const messageSchema = new mongoose.Schema({
  chatId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Chat',
    required: true
  },
  senderId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  text: {
    type: String,
    required: true
  },
  type: {
    type: String,
    enum: ['TEXT', 'IMAGE'],
    default: 'TEXT'
  },
  mediaUrl: {
    type: String,
    default: null
  },
  mediaType: {
    type: String,
    default: null
  },
  replyToMessageId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Message',
    default: null
  },
  replyToText: {
    type: String,
    default: null
  },
  replyToSenderId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    default: null
  },
  status: {
    type: String,
    enum: ['SENDING', 'SENT', 'DELIVERED', 'READ', 'FAILED'],
    default: 'SENT'
  },
  isEdited: {
    type: Boolean,
    default: false
  },
  editedAt: {
    type: Date,
    default: null
  },
  isDeleted: {
    type: Boolean,
    default: false
  }
}, {
  timestamps: true
});

// Index for faster queries
messageSchema.index({ chatId: 1, createdAt: -1 });

module.exports = mongoose.model('Message', messageSchema);

File: ChatFlow/app/src/main/java/com/chatflow/app/domain/model/Message.kt
package com.chatflow.app.domain.model

enum class MessageType {
    TEXT,
    IMAGE,
    VIDEO,
    AUDIO,
    DOCUMENT,
    LOCATION
}

enum class MessageStatus {
    SENDING,
    SENT,
    DELIVERED,
    READ,
    FAILED
}

data class Message(
    val messageId: String,
    val senderId: String,
    val text: String,
    val type: MessageType,
    val mediaUrl: String?,
    val mediaType: String?,
    val mediaDuration: Long?,
    val replyToMessageId: String?,
    val replyToText: String?,
    val replyToSenderId: String?,
    val timestamp: Long,
    val status: MessageStatus,
    val isEdited: Boolean,
    val editedAt: Long?,
    val isDeleted: Boolean
)
